include/fi_mem.h

```diff
@@ -69,24 +69,32 @@ static inline void *mem_dup(const void *src, size_t size)
  */
 #define FREESTACK_EMPTY	NULL
 
+#define FREESTACK_HEADER 					\
+	size_t		size;					\
+	void		*next;					\
+
 #define freestack_isempty(fs)	((fs)->next == FREESTACK_EMPTY)
 #define freestack_push(fs, p)					\
 {								\
 	*(void **) p = (fs)->next;				\
 	(fs)->next = p;						\
 }
-#define freestack_pop(fs)					\
-	({							\
-		void *p = (fs)->next;				\
-		assert(!freestack_isempty(fs));			\
-		(fs)->next = *((void **) p);			\
-		p;						\
-	})
+#define freestack_pop(fs) freestack_pop_impl(fs, fs->next)
+
+static inline void* freestack_pop_impl(void *fs, void *fs_next)
+{
+	struct {
+		FREESTACK_HEADER;
+	} *_fs = fs;
+	void *p = fs_next;
```

Comment:
Why use a `p` pointer when you already have a variable with the same type and same value (`fs_next`)?

Seems like this might be cleaner if, instead of using a macro that expands to a set of fields, we bundled those fields into a proper struct and passed a pointer to this struct instead of a pointer to the outer structure.  The downside is that it would require updating the other freestack macros.